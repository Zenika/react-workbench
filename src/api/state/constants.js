const { COMPONENT_ABSOLUTE_PATH_DIR, PROJECT_CONFIG_DIR } = require('../../constants')

const NAME = 'state'
const DATABASE_PATH = `${COMPONENT_ABSOLUTE_PATH_DIR}/${PROJECT_CONFIG_DIR}/${NAME}.json`

module.exports = {
  NAME,
  DATABASE_PATH,
}
