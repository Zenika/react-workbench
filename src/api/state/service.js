// TODO : these two lines are duplicate everywhere
const Promise = require('bluebird')
const fs = Promise.promisifyAll(require('fs'))
const { COMPONENT_CONFIG_DIR } = require('../../constants')
const { DATABASE_PATH } = require('./constants')

// TODO : move "database" controle so it can be reused
const create = async (state) => {
  // TODO : flow control of state ?

  // read previous states
  let content = []
  try {
    content = JSON.parse(await fs.readFileAsync(DATABASE_PATH))
  } catch (ex) {
    if (ex.errno !== -2) throw ex // -2 is file not found
  }

  // add new state
  content = content.concat(state)

  // create configuration directory
  try {
    await fs.mkdirAsync(COMPONENT_CONFIG_DIR)
  } catch (ex) {
    if (ex.errno !== -17) throw ex // -17 is directory exist
  }

  // write states
  await fs.writeFileAsync(DATABASE_PATH, JSON.stringify(content))
}

module.exports = {
  create,
}
