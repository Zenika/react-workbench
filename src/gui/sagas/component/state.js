import { put, select } from 'redux-saga/effects'
import { state as uiStore } from '../../redux/ui'
import { state as store } from '../../redux/component'

export function* setUiState(state) {
  yield put(uiStore.set(JSON.stringify(state, null, 2)))
}

export function* setComponentState() {
  const state = yield select(uiStore.get())
  yield put(store.set(JSON.parse(state)))
}
