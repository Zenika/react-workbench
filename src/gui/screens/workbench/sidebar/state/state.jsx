import React from 'react'
import PropTypes from 'prop-types'
import { onlyUpdateForPropTypes } from 'recompose'
import { merge } from 'glamor'
import styles from './state.styles'

const State = ({ style, className, state, onChange, setState }) => {
  return (
    <div style={style} className={merge(styles.layout, className)}>
      <button onClick={setState}>set</button>
      <textarea className={styles.textarea} value={state} onChange={onChange} />
    </div>
  )
}

State.propTypes = {
  style: PropTypes.object,
  className: PropTypes.string,
  state: PropTypes.object,
  onChange: PropTypes.func,
  setState: PropTypes.func,
}

State.defaultProps = {
  style: {},
  className: '',
  state: undefined,
  onChange: undefined,
  setState: undefined,
}

export default onlyUpdateForPropTypes(State)
