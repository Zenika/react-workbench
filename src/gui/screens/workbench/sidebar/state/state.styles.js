import { css } from 'glamor'

const layout = css({
  display: 'flex',
  flexDirection: 'column',
  height: 'calc(100% - 3em)',
})

const textarea = css({
  flex: '1',
})

export default { layout, textarea }
