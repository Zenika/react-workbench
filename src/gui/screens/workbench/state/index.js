export { default } from './state.container'
