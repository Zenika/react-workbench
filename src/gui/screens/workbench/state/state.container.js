import { connect } from 'react-redux'
import { state as stateStore } from '../../../redux/ui'
import Component from './state'

const mapState = state => ({ state: stateStore.get()(state) })

const mapDispatch = dispatch => ({
  onChange: e => dispatch(stateStore.set(e.target.value)),
  setState: () => dispatch({ type: 'UI_CLICK_SET_STATE' }),
})

export default connect(mapState, mapDispatch)(Component)
