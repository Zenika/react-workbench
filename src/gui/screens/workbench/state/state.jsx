import React from 'react'
import PropTypes from 'prop-types'
import { onlyUpdateForPropTypes } from 'recompose'

const State = ({ style, className, state, onChange, setState }) => {
  return (
    <div style={style} className={className}>
      <textarea value={state} onChange={onChange} />
      <button onClick={setState}>set</button>
    </div>
  )
}

State.propTypes = {
  style: PropTypes.object,
  className: PropTypes.string,
  state: PropTypes.object,
  onChange: PropTypes.func,
  setState: PropTypes.func,
}

State.defaultProps = {
  style: {},
  className: '',
  state: undefined,
  onChange: undefined,
  setState: undefined,
}

export default onlyUpdateForPropTypes(State)
