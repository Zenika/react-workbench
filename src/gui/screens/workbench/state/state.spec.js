/* eslint-env jest */
import snap from 'misc/test/snap'
import state from './index' // eslint-disable-line
import Graphical from './state'
import { mapState, mapDispatch } from './state.container'

describe('workbench/state', () => {
  describe('graphical', () => {
    const snapshot = snap(Graphical)

    // FIXME : merge on className seems not to work
    it('should print props', snapshot({ style: { backgroundColor: 'red' }, className: 'custom', onChange: () => {}, setState: () => {} }))
    it('should have a default behaviour', snapshot({}))
  })

  describe('container', () => {
    it('should map state', () => {
      const props = mapState({
        ui: {
          state: 'some state',
        },
      })

      expect(props).toMatchSnapshot()
    })

    it('should map dispatch', () => {
      const dispatch = jest.fn()

      // object
      const props = mapDispatch(dispatch)
      expect(props).toMatchSnapshot()

      // events
      props.onChange({ target: { value: 'input value' } })
      props.setState()

      expect(dispatch.mock.calls).toMatchSnapshot()
    })
  })
})
