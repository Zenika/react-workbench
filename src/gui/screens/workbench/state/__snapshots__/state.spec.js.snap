// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`workbench/state container should map dispatch 1`] = `
Object {
  "onChange": [Function],
  "setState": [Function],
}
`;

exports[`workbench/state container should map dispatch 2`] = `
Array [
  Array [
    Object {
      "payload": "input value",
      "type": "@trampss/SET_UISTATE",
    },
  ],
  Array [
    Object {
      "type": "UI_CLICK_SET_STATE",
    },
  ],
]
`;

exports[`workbench/state container should map state 1`] = `
Object {
  "state": "some state",
}
`;

exports[`workbench/state graphical should have a default behaviour 1`] = `
<div
  className={
    Object {
      "data-css-10lnnxw": "",
    }
  }
  style={Object {}}
>
  <button
    onClick={undefined}
  >
    set
  </button>
  <textarea
    className={
      Object {
        "data-css-n9fe5c": "",
      }
    }
    onChange={undefined}
    value={undefined}
  />
</div>
`;

exports[`workbench/state graphical should print props 1`] = `
<div
  className={
    Object {
      "data-css-19n021x": "",
    }
  }
  style={
    Object {
      "backgroundColor": "red",
    }
  }
>
  <button
    onClick={[Function]}
  >
    set
  </button>
  <textarea
    className={
      Object {
        "data-css-n9fe5c": "",
      }
    }
    onChange={[Function]}
    value={undefined}
  />
</div>
`;
